Major upgrade: Changes documented

tractorcow opened the pull request:
@Firesphere would love to discuss this with you before merging; See you on slack maybe?

Overview of changes in general order of most significant to least

- Return type of all validate / create mutations has been consolidated and normalised. The standard response is now a `MemberToken` type with these fields:
  - Member (the member)
  - Token (string token) 
  - Valid (bool valid flag)
  - Status (enum status code of the token)
  - Code (int HTTP code that matches the status)
  - Message (string message of humanised status code)
- Multiple authenticator tokens per user; Allows multiple devices to have their own tokens.
- Anonymous authentication is now shifted up into a custom Authenticator class; Support injectable authenticators into createToken mutation. Removed `anonymous_allowed`
- When logging in anonymously, create a `real` member object to better integrate with silverstripe security logging / auditing practices.
- Bump minimum dependency to php 7.1: strict_types are everywhere too.
- Updated tests and documentation of course
- Support multiple schemas (graphql 3.0 and above).
- Relative path supported for JWT tokens
- Updated renewal process to support separate token / refresh expiration dates
- I gave every file my Love. Green ticks everywhere.

@Firesphere if you merge, I suggest branching master -> 1 first, as this is a new 2.x major change.

tractorcow committed:
Commit refactored work

tractorcow committed:
Anonymous authenticator
Fix and upgrade tests
Upgrade documentation
Bump minimum dependency to 7.1

tractorcow committed:
I put Message on the wrong type

tractorcow committed:
Bump major version

Firesphere commented:
Could you make a minor change (I don't know, docs or something?) to ping CircleCI to run? I don't know why it didn't, but I've reset the PR settings back to default so it should work again.

Firesphere commented on src/Authentication/JWTAuthenticationHandler.php:
```diff
@@ -1,61 +1,50 @@
 
     /**
      * @param HTTPRequest $request
      * @return null|Member
```
Return can be removed if return types are in PHP7 style

tractorcow replied:
Maybe, unless the return type is one of several types (e.g. `string|int|bool`) or if the return type has a comment. I'd rather just keep it consistently than go through and only have them on some methods.

I might wait until psr-5 is finalised to make a call on this https://github.com/php-fig/fig-standards/blob/master/proposed/phpdoc.md

Firesphere replied:
Yep, fair enough :)

Firesphere commented on src/Authentication/JWTAuthenticationHandler.php:
```diff
@@ -67,32 +56,28 @@ public function authenticateRequest(HTTPRequest $request)
      *
-     * @param Member $member
-     * @param bool $persistent
+     * @param Member           $member
```
Matter of preference I guess? I'm not too fuzzed, but I prefer tight aligned for docblocks.

Not an issue for me though, just saying :)

tractorcow replied:
Normally I'm the one that fussed. Sorry I'll fix those. :)

Firesphere commented on src/Authentication/AnonymousUserAuthenticator.php:
```diff
@@ -0,0 +1,115 @@
+     * @return Member Found member, regardless of successful login
+     * @throws ValidationException
+     */
+    protected function authenticateMember($data, ValidationResult &$result = null, Member $member = null): Member
```
Do you see an option to use a sort of `AnonymousMember` IdentityStore? I think it would be cleaner in the long run

tractorcow replied:
Since the persistence has moved out of the application to the client (JWT for the "session" is stored on the client) I'm not sure it really makes sense in this context.

IdentityStore requires a member is passed in to it, which doesn't really let it incapsulate the behaviour I'm trying to build in here (i.e. acts as a factory for the anonymous member itself).

The rest of the persistence is handled by the mutations / queries as well.

Maybe we could do both in the future? You might need to take the initiative here to show me what you mean I think. :)

Firesphere commented on src/Authentication/AnonymousUserAuthenticator.php:
```diff
@@ -0,0 +1,115 @@
+    protected function authenticateMember($data, ValidationResult &$result = null, Member $member = null): Member
+    {
+        // Get user, or create if not exists
+        $member = $this->getOrCreateAnonymousMember();
```
Would/could this possibly be a service instead of a method on `$this`? I'm aware my original wasn't that clean either, but might as well do it entirely right from this refactor

tractorcow replied:
Yes a factory wouldn't hurt.

tractorcow replied:
Pushed up.

Firesphere commented on src/Authentication/AnonymousUserAuthenticator.php:
```diff
@@ -0,0 +1,115 @@
+
+        // Create new member
+        $member = Member::create();
+        $member->{$identifierField} = $username;
```
For cleaner readability, could you use the __set() method here?

Firesphere commented on src/Authentication/AnonymousUserAuthenticator.php:
```diff
@@ -0,0 +1,115 @@
+        $member = Member::create();
+        $member->{$identifierField} = $username;
+        $member->update($fields);
+        $this->extend('updateCreatedAnonymousMember', $member);
```
Do you think there is any need for this extension? It could create a security hole if the engineer extending isn't experienced enough

tractorcow replied:
Dropping it since we'll have a factory instead. We can just override the factory with yml.

Firesphere replied:
Perfect

Firesphere commented on src/Authentication/JWTAuthenticationHandler.php:
```diff
@@ -1,61 +1,50 @@
-<?php
+<?php declare(strict_types=1);
```
YES!

Firesphere commented on src/Authentication/JWTAuthenticator.php:
```diff
@@ -25,201 +34,252 @@
+    private function getPublicKey(): Key
     {
         $signerKey = Environment::getEnv('JWT_SIGNER_KEY');
+        $signerPath = PathResolver::resolve($signerKey);
```
Nice extraction to PathResolver!

Firesphere commented on src/Authentication/JWTAuthenticator.php:
```diff
@@ -25,201 +34,252 @@
+     * @throws LogicException
      */
-    private function setKeys()
+    private function getPublicKey(): Key
```
I think it would be good to unify this method and simply use `getKey($type)` where type defines if it's public or private. It would reduce code duplication IMO

tractorcow replied:
Yeah what did I do here. Normally I'm the one who complains about this...

Although public and private keys are different; I can probably still clean this code up.

tractorcow replied:
Result is nice; I'll push it up soon.

I did add a getKeyType() to consolidate the if / else logic. getPublicKey() / getPrivateKey() just call the consolidated method.

Firesphere commented on src/Authentication/JWTAuthenticator.php:
```diff
@@ -25,201 +34,252 @@
 
-        return $this->validateToken($token, $request, $result);
+        /** @var JWTRecord $record */
+        list($record, $status) = $this->validateToken($token, $request);
```
I so much hate and love the `list()` record! :D I use it a lot but I still feel dirty for doing it.

tractorcow replied:
Wish PHP had object destructuring hah.

Firesphere commented on src/Authentication/JWTAuthenticator.php:
```diff
@@ -25,201 +34,252 @@
             ->setId($uniqueID, true)
             // Configures the time that the token was issue (iat claim)
-            ->setIssuedAt(time())
+            ->setIssuedAt($now)
```
This should be configurable actually, neither my original time() nor $now is correct to me, it should take in to account any options to set it in the future from config I'd think

tractorcow replied:
When do we want to set a different issued at date? Maybe another discussion huh.

Firesphere replied:
Yeah, probably better off as a separate issue

Firesphere replied:
Issue: https://github.com/Firesphere/silverstripe-graphql-jwt/issues/25

Firesphere commented on src/Authentication/JWTAuthenticator.php:
```diff
@@ -25,201 +34,252 @@
+            // Set renew expiration
+            ->set('rexp', $now + $config->get('nbf_refresh_expiration'))
+            // Configures a new claim, called "rid"
+            ->set('rid', $record->ID)
```
👍 Gah, I've been wanting to fix this up for ages! Thanks!

Firesphere commented on src/Extensions/MemberExtension.php:
```diff
@@ -68,4 +64,17 @@ public function getJWTData()
+     *
+     * @return Member
+     */
+    public function DestroyAuthTokens(): Member
```
Not PSR compliant to start with a capital

tractorcow replied:
That was authully careless of me.

Firesphere commented on src/Extensions/MemberExtension.php:
```diff
@@ -68,4 +64,17 @@ public function getJWTData()
 
         return Convert::raw2json($data);
```
raw2json has been superseeded with direct json_encode, without calling `Convert` class (unrelated but related, might as well fix it up straight away)

Firesphere commented on src/Helpers/MemberTokenGenerator.php:
```diff
@@ -0,0 +1,76 @@
+    protected function generateResponse(string $status, Member $member = null, string $token = null): array
+    {
+        // Success response
+        if ($status == TokenStatusEnum::STATUS_OK) {
```
This if/else construction could be replaced with a direct return? Or the else above the if and skipping `else`

tractorcow replied:
There's an extension hook that would need to be duplicated then.

Firesphere replied:
Cool, leave as is then :)

Firesphere commented on src/Helpers/PathResolver.php:
```diff
@@ -0,0 +1,22 @@
+
+namespace Firesphere\GraphQLJWT\Helpers;
+
+class PathResolver
```
This could/should be a core trait or class I'd think

Firesphere commented on src/Helpers/RequiresConfig.php:
```diff
@@ -0,0 +1,29 @@
+     * @throws LogicException Error if environment variable is required, but not configured
+     * @return mixed
+     */
+    public static function getEnv(string $key, $default = null)
```
As above, a type here might be convenient for future use

Firesphere commented on src/Model/JWTRecord.php:
```diff
@@ -0,0 +1,25 @@
+ */
+class JWTRecord extends DataObject
+{
+    private static $table_name = 'Firesphere_JWTRecord';
```
Any specific reason for the prefix?

tractorcow replied:
Not really. :D

Firesphere commented on src/Model/JWTRecord.php:
```diff
@@ -0,0 +1,25 @@
+    private static $table_name = 'Firesphere_JWTRecord';
+
+    private static $db = [
+        'UID'       => 'Varchar(255)',
```
UID sounds like it could use an index declaration for DB speed

Firesphere reviewed (changes requested):
Only halfway through the review, but here are some thoughts :)

Firesphere commented:
Could you have a look at the CodeClimate and Scrutinizer issues? They're not super important or anything, but it's not a bad thing to keep an eye on :)

Firesphere commented:
Could you have a look at the CodeClimate failures? They're not of high importance, but good to keep them low.
Also, where did my previous comment go??!

Firesphere commented on .editorconfig:
```diff
@@ -0,0 +1,26 @@
+# For more information about the properties used in
```
Any reason for including this file? I don't mind, but I usually only have it in a project root, not a module?

Firesphere commented on src/Authentication/AnonymousUserAuthenticator.php:
```diff
@@ -0,0 +1,115 @@
+    {
+        // Only applies to request for anonymous user specifically
+        $email = $data['Email'] ?? null;
+        if ($email !== static::config()->get('anonymous_username')) {
```
So this constrains the anonymous username to 'anonymous'?

Firesphere commented on src/Authentication/JWTAuthenticationHandler.php:
```diff
@@ -67,32 +56,28 @@ public function authenticateRequest(HTTPRequest $request)
+        if ($member) {
+            $member->DestroyAuthTokens();
         }
-        // Empty the current user and pray to god it's not valid anywhere else anymore :)
```
Aww, you removed my funny comment! 😂

Firesphere commented on src/Authentication/JWTAuthenticator.php:
```diff
@@ -25,201 +34,252 @@
+        try {
+            $parsedToken = $parser->parse($token);
+        } catch (Exception $ex) {
+            // Un-parsable tokens are invalid
```
Could you add some logging here?

Firesphere commented on src/Extensions/MemberExtension.php:
```diff
@@ -1,61 +1,57 @@
         $identifier = Member::config()->get('unique_identifier_field');
         $extraFields = Member::config()->get('jwt_subject_fields');
 
+        $data->type = 'member';
```
Could you make this `ClassLoader::shortClassName($this->owner->ClassName);`? I think it's a tad cleaner, allowing for extended members to have their own type?

Firesphere commented on src/Mutations/CreateTokenMutationCreator.php:
```diff
@@ -1,83 +1,137 @@
+                return $member;
+            }
         }
 
```
Would it be cleaner to return the result no matter what? setting an error that no valid member has been found

Firesphere commented on src/Mutations/CreateTokenMutationCreator.php:
```diff
@@ -1,83 +1,137 @@
+    protected function getLoginAuthenticators(): Generator
+    {
+        // Check injected authenticators
+        yield from $this->getCustomAuthenticators();
```
Any specific reason for the `yield`? I'm too unexperienced with yield to properly know what it's useful for (besides a bit of async stuff)

Firesphere commented on src/Queries/ValidateTokenQueryCreator.php:
```diff
@@ -1,73 +1,70 @@
-        return ['Valid' => $result->isValid(), 'Message' => implode('; ', $msg), 'Code' => $code];
+        /** @var JWTRecord $record */
+        list($record, $status) = $authenticator->validateToken($token, $request);
+        $member = $status === TokenStatusEnum::STATUS_OK ? $record->Member() : null;
```
This is not entirely readable, maybe add brackets or make it a full if?

Firesphere commented on src/Types/TokenStatusEnum.php:
```diff
@@ -0,0 +1,81 @@
+
+use GraphQL\Type\Definition\EnumType;
+
+class TokenStatusEnum extends EnumType
```
Could you add some docblock as to what this class exactly is for?

Firesphere reviewed (changes requested):
I've skipped over the tests so far, I'll need some time next week to properly actually manually test it :)

Firesphere commented on src/Types/TokenStatusEnum.php:
```diff
@@ -0,0 +1,81 @@
+
+    public function __construct()
+    {
+        $values = [
```
Could you extract this in to a static from config? It doesn't look like it needs to be written fully in the construct?

Firesphere commented on src/Authentication/JWTAuthenticator.php:
```diff
@@ -25,201 +34,252 @@
+    public function authenticate(array $data, HTTPRequest $request, ValidationResult &$result = null): ?Member
     {
         if (!$result) {
             $result = new ValidationResult();
```
Should be `ValidationResult::create()`

Firesphere commented on src/Authentication/JWTAuthenticator.php:
```diff
@@ -25,201 +34,252 @@
-        $verified = $parsedToken->verify($this->signer, $this->publicKey);
+        // Get validator for this token
+        $now = DBDatetime::now()->getTimestamp();
+        $validator = new ValidationData();
```
`ValidationData::create()`?

tractorcow replied:
It's  not a silverstripe class. Can't use that.

Firesphere replied:
Okay :)

Firesphere commented on src/Mutations/CreateTokenMutationCreator.php:
```diff
@@ -1,83 +1,137 @@
+    {
+        // Login with authenticators
+        foreach ($this->getLoginAuthenticators() as $authenticator) {
+            $result = new ValidationResult();
```
`ValidationResult::create()`?

Firesphere commented:
Could you run PHP-CS-Fixer to fix the PSR-2 violations?

tractorcow commented:
Great feedback so far. I'll get back later this week. :D

tractorcow committed:
Restore tight docblocks

tractorcow committed:
Move anonymous user generation to a factory

tractorcow committed:
Refactor key generation code

tractorcow committed:
Remove unused classes
Merge single-use traits into parent classes for simplicity
Add DB index
Cleanup phpcs violations

tractorcow commented:
@Firesphere I've done a quick as guts refactor. All work is in separate commits for your review.

traits that were just used once I've inlined.

Removed unused JWTException.

I hope the getKeys() code suits you. It was kind of weird before.

I haven't done much except run the unit tests and phpcs. I'll need to come back and test more later.

Firesphere commented:
Still confused why CircleCI won't run... but that refactor looks great!

Just a quick check, does this incorporate a possible fix fo #22 ?

Firesphere commented:
Ah, this is why the tests don't run:
```
    branches:
      only:
        - master
        - develop
        - /feature.*/
```
in .circleci/config.yml

Mind removing those lines for me? Otherwise, you'd have to rebase for such a simple change ;)

tractorcow commented:
#22 would probably require a new pull request. I don't really have time at the moment I'm afraid.

tractorcow committed:
Remove branch limits to circleci

tractorcow commented:
Removed those lines.

tractorcow commented:
Oh I love the codeclimate output. :)

tractorcow commented:
@Firesphere now that it's working, do you mind doing a bit of triaging on those codeclimate issues and tell me which I need to fix? I'd fix them all if you say, although I question whether the recommendations are ideal.

Firesphere commented:
I'll try to have a look today, CodeClimate reports aren't that big a deal for me usually, as more often than not it's unrelated stuff (e.g. SilverStripe's use of statics ;) )

tractorcow commented:
Yeah there's stuff there I can fix. I don't know why codeclimate hates having multiple returns (I prefer that over more nested code structures). I can do some work end of day to tidy it up though.

Firesphere commented on src/Authentication/JWTAuthenticator.php:
```diff
@@ -155,71 +271,104 @@ public function generateToken(Member $member)
      * @throws BadMethodCallException
      */
-    private function validateToken($token, $request, &$result)
+    public function validateToken(string $token, HTTPrequest $request): array
```
Actual valid issue from CodeClimate: It's quite a long method:
https://codeclimate.com/github/Firesphere/silverstripe-graphql-jwt/src/Authentication/JWTAuthenticator.php/source?from_sha=2a45ae2a5722e3685f0c5db06f46dbd3b6eb80d2&to_sha=8fd6b877d78de4d520a9eb566a757e017bbb8ea5
And has a lot of return statements.

Could you reduce the returns (maybe not an early return each time?) and maybe refactor the ValidationData part in to a separate getter?

tractorcow replied:
I've broken this into a bunch of methods. There are still a few early returns, but a few fewer than before.

Firesphere commented on src/Helpers/MemberTokenGenerator.php:
```diff
@@ -0,0 +1,76 @@
+    protected function generateResponse(string $status, Member $member = null, string $token = null): array
+    {
+        // Success response
+        if ($status === TokenStatusEnum::STATUS_OK) {
```
Could you change this to
```
            $response = [
                'Valid'   => false,
                'Member'  => null,
                'Token'   => $token,
                'Status'  => $status,
                'Code'    => 401,
                'Message' => $this->getErrorMessage($status),
            ];
```
followed by the if (omitting the else). As per CodeClimate suggestion:
https://codeclimate.com/github/Firesphere/silverstripe-graphql-jwt/src/Helpers/MemberTokenGenerator.php/source?from_sha=2a45ae2a&to_sha=8fd6b877

tractorcow replied:
![image](https://user-images.githubusercontent.com/936064/56869933-ddf68680-6a5b-11e9-9df9-4768f57dfcbe.png)

What?

tractorcow replied:
I'll just switch it to a single assignment, and move the conditionals to the line items.

Really weird coding conventions I must exclaim...

tractorcow replied:
I don't use an if at all now. :D

tractorcow replied:
Apparently this is a thing http://elseisnevernecessary.com

Firesphere reviewed (changes requested):
Just 2 things from CodeClimate, and could you have a general look at
https://codeclimate.com/github/Firesphere/silverstripe-graphql-jwt/issues?category%5B%5D=complexity&status%5B%5D=&status%5B%5D=open&status%5B%5D=confirmed

To see if there's anything worth picking up beyond the two mentioned? Most are not really that important.

Firesphere commented:
I don't see any major problems besides the two mentioned for maintainability. I'll give everything a run for its money somewhere this weekend :) And possibly push some minor doc changes or personal preferences :)

Firesphere commented:
> Oh I love the codeclimate output. :)

I think you mean CodeCov, but yeah, same, love that output back as a comment! :D

Firesphere commented:
Ow, could you please comment on the open issues that are resolved in this PR as such with reference? Would make it easier for history (and for me to close them :) )

Firesphere commented:
> @Firesphere if you merge, I suggest branching master -> 1 first, as this is a new 2.x major change.

I think I'm going to go with a standard tagging, way too okay with these changes generally to even bother keeping the old start I made with 2.0 active. Tags should suffice and I can always branch off the tag later if needed :)

tractorcow committed:
Break up large method to appease codeclimate

tractorcow commented:
@Firesphere please check my latest commit. Sorry I didn't get the code perfect, but it is an improvement on the major style issues. I'm just time constrained sorry.

tractorcow commented:
Hold on, don't merge! I've run into this problem with user-agent validation.

`User-Agent: Mozilla/5.0 (Macintosh; Intel Mac OS X 10_13_6) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/74.0.3729.108 Safari/537.36`

Means a weekend-update (or mid-day update) to chrome will break all your sessions. `74.0.3729.107` to `74.0.3729.108` means you need to re-login. =(

tractorcow committed:
Remove breakage on user-agent changing

tractorcow commented:
Fixed the thing I was worried about @Firesphere :)

Firesphere commented:
Note to self: Look at this tonight and finally merge it in!

tractorcow commented:
yeah that'd be great. :)

ec8or commented:
Would be awesome! ;)
